hack/generated/pkg/genruntime/resource_reference.go

```diff
@@ -67,6 +70,24 @@ func (ref ResourceReference) String() string {
 	return fmt.Sprintf("Group: %q, Kind: %q, Namespace: %q, Name: %q, ARMID: %q", ref.Group, ref.Kind, ref.Namespace, ref.Name, ref.ARMID)
 }
 
+// TODO: We wouldn't need this if controller-gen supported DUs or OneOf better, see: https://github.com/kubernetes-sigs/controller-tools/issues/461
+// Validate validates the ResourceReference to ensure that it is structurally valid.
+func (ref ResourceReference) Validate() error {
+	if ref.ARMID == "" && ref.Namespace == "" && ref.Name == "" && ref.Group == "" && ref.Kind == "" {
+		return errors.Errorf("at least one of [ARMID] or [Group, Kind, Namespace, Name] must be set for ResourceReference")
+	}
+
+	if ref.ARMID != "" && (ref.Namespace != "" || ref.Name != "" || ref.Group != "" || ref.Kind != "") {
```

Comment:
If we did this, it would end up being:

`if ref.ARMID != "" && !(ref.ARMID != "" && ref.Namespace == "" && ref.Name == "" && ref.Group == "" && ref.Kind == "")` 
Which simplifies to:
 `if ref.ARMID != "" && (ref.ARMID == "" || ref.Namespace != "" || ref.Name != "" || ref.Group != "" || ref.Kind != "")`

Which isn't the same -- because it has `if ref.ARMID != "" && ref.ARMID == ""` which can never be true.


I could maybe make some nonexported functions that were building blocks?
```
func (ref ResourceReference) hasAnyKubernetesReferenceFieldsSet() bool {
	return ref.Namespace != "" || ref.Name != "" || ref.Group != "" || ref.Kind != ""
}
```

```
func (ref ResourceReference) hasARMIDFieldSet() bool {
	return ref.ARMID != ""
}
```

?

Reply:
Bah I dropped the parenthesis when I should've have. I think you're right. Can fix this